Fix untyped malloc() calls in JavaScriptCore and WebCore

ddkilzer opened the pull request:
#### eaa04bef99bc2de35caf3d664950ef4cba450379
<pre>
Fix untyped malloc() calls in JavaScriptCore and WebCore
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=301118">https://bugs.webkit.org/show_bug.cgi?id=301118</a>&gt;
&lt;<a href="https://rdar.apple.com/163059754">rdar://163059754</a>&gt;

Reviewed by Geoff Garen.

When compiling with type-aware malloc enabled, the compiler has many
hints from which type information may be deduced:
- C-style casts
- reinterpret_cast&lt;&gt;()
- static_cast&lt;&gt;()
- Multiplying a variable by sizeof(TYPE_NAME)

The fix is to find a way to introduce these mechanisms, or to use
WTF_TYPE_AWARE_MALLOC_FUNCTION() when C++ template types aren&apos;t
available.

* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::dumpJITMemory):
- Fix untyped malloc() by switching to LazyNeverDestroyed&lt;std::array&gt;.
- Also fix remaining unsafe buffer issues by adopting std::span.

* Source/WTF/wtf/MallocSpan.h:
- Provide specialization for MallocSpan&lt;T, SystemMalloc&gt; that inherits
  from MallocSpan&lt;T, SystemMallocBase&lt;T&gt;&gt; so that &lt;typename T&gt; is
  available when using SystemMallocBase in order to provide type hints
  to malloc() functions via static_cast&lt;T*&gt;().  We use this
  specialization so that other classes used with WTF::MallocSpan are not
  required to adopt a &lt;typename T&gt; template parameter.
* Source/WTF/wtf/PlatformHave.h:
- Change definition of HAVE_TYPE_AWARE_MALLOC and
  WTF_TYPE_AWARE_MALLOC_FUNCTION() to use OS versions.  An enablement
  check is added to SystemMalloc.h where &lt;malloc/malloc.h&gt; may be
  included without affecting preprocesesing of sandbox profiles
  (*.sb.in).
* Source/WTF/wtf/SystemMalloc.h:
- Add enablement checks that _MALLOC_TYPE_ENABLED and
  HAVE_TYPE_AWARE_MALLOC are both enabled at the same time.
- Rename WTF::SystemMalloc to WTF::SystemMallocBase for SystemMalloc
  template specialization in MallocSpan.h (see above).
(WTF::SystemMallocBase::malloc):
(WTF::SystemMallocBase::tryMalloc):
(WTF::SystemMallocBase::zeroedMalloc):
(WTF::SystemMallocBase::tryZeroedMalloc):
(WTF::SystemMallocBase::realloc):
(WTF::SystemMallocBase::tryRealloc):
- Add static_cast&lt;T*&gt;() to provide type hint to malloc() and friends,
  then change the return type from void* to T*.

(WTF::SystemMalloc::malloc): Rename.
(WTF::SystemMalloc::tryMalloc): Rename.
(WTF::SystemMalloc::zeroedMalloc): Rename.
(WTF::SystemMalloc::tryZeroedMalloc): Rename.
(WTF::SystemMalloc::realloc): Rename.
(WTF::SystemMalloc::tryRealloc): Rename.
(WTF::SystemMalloc::free): Rename.
(WTF::SystemMalloc::nextCapacity): Rename.
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::xmlMallocHelper):
(WebCore::xmlFreeHelper):
(WebCore::XMLMalloc::malloc):
- When HAVE(TYPE_AWARE_MALLOC) is enabled, just make XMLMalloc an alias
  of WTF::SystemMalloc since that&apos;s effectively what we were doing
  before.  This cleans up typed malloc support for this code.
(WebCore::XMLMalloc::WTF_TYPE_AWARE_MALLOC_FUNCTION): Deleted.

Canonical link: <a href="https://commits.webkit.org/302125@main">https://commits.webkit.org/302125@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/972d090c043f007522ad5d84f3c11ac50d3aef97

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128044 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135414 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79554 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9a6812ee-24a9-418a-a0fa-501f59fcca00) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129916 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/200 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97478 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65372 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ef7f6012-6c81-4229-b2d3-7558c75d98c6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130992 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/140 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114704 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78045 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2f008188-4957-469d-bfa8-22de0591a317) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/136 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32813 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78723 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/120090 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108505 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33294 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137902 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126505 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/183 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/190 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106008 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/208 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111053 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105744 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/139 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29609 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52362 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20015 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/229 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/61695 "") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159526 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/151 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39823 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/226 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/189 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
```diff
@@ -115,33 +115,26 @@ static inline bool shouldRenderInXMLTreeViewerMode(Document& document)
-static void* xmlMallocHelper(size_t size, malloc_type_id_t typeID)
-{
-    return malloc_type_malloc(size, typeID);
-}
+using XMLMalloc = WTF::SystemMalloc;
 #else
```
@achristensen07 This fix resolves some of the ugliness you had concerns about in PR https://github.com/WebKit/WebKit/pull/48300.

ddkilzer commented on Source/WTF/wtf/SystemMalloc.h:
```diff
@@ -25,30 +25,57 @@
+#define WTF_TYPE_AWARE_MALLOC_FUNCTION(function, index) _MALLOC_TYPED(function, index)
+#else
+#define WTF_TYPE_AWARE_MALLOC_FUNCTION(function, index)
+#endif
```
This is a check-webkit-style violation, but I'm not sure how best to resolve it.

We want to ensure that `<malloc/malloc.h>` is included wherever `HAVE(TYPE_AWARE_MALLOC)` or `WTF_TYPE_AWARE_MALLOC_FUNCTION()` is used, but we can't do that in `PlatformHave.h` itself since that breaks preprocessing of sandbox profiles (`*.sb.in`).

```
Source/WTF/wtf/SystemMalloc.h:43: error: [build/version_check] Misplaced OS version check. Please use a named macro in one of headers in the wtf/Platform.h suite of files or an appropriate internal file.
Source/WTF/wtf/SystemMalloc.h:44: error: [build/version_check] Misplaced OS version check. Please use a named macro in one of headers in the wtf/Platform.h suite of files or an appropriate internal file.
Source/WTF/wtf/SystemMalloc.h:45: error: [build/version_check] Misplaced OS version check. Please use a named macro in one of headers in the wtf/Platform.h suite of files or an appropriate internal file.
Source/WTF/wtf/SystemMalloc.h:46: error: [build/version_check] Misplaced OS version check. Please use a named macro in one of headers in the wtf/Platform.h suite of files or an appropriate internal file.
Source/WTF/wtf/SystemMalloc.h:47: error: [build/version_check] Misplaced OS version check. Please use a named macro in one of headers in the wtf/Platform.h suite of files or an appropriate internal file.
Total errors found: 5 in 4 files
```

ddkilzer replied:
I found a way around this by defining `HAVE(TYPE_AWARE_MALLOC)` in `PlatformHave.h`, but adding sanity checks to `SystemMalloc.h`.

Constellation commented on Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
```diff
@@ -1372,7 +1372,8 @@ void dumpJITMemory(const void* dst, const void* src, size_t size)
     std::call_once(once, [] {
-        buffer = std::bit_cast<uint8_t*>(malloc(bufferSize));
+        // FIXME: <rdar://162872180> std::bit_cast does not provide a hint to malloc().
+        buffer = std::bit_cast<uint8_t*>(malloc(bufferSize * sizeof(*buffer)));
```
Actually, this is not intentionally using `malloc`. So `LazyNeverDestroyed<Vector<uint8_t>>` is fine ultimately. But as it is only used for CPU data collection (very specific use case for data collection), just changing it to this form is also fine.

ddkilzer replied:
Actually going to switch to `LazyNeverDestroyed<std::array<uint8_t, bufferSize>>` since the buffer is fixed-size.  :)

geoffreygaren commented on Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
```diff
@@ -1387,10 +1387,10 @@ void dumpJITMemory(const void* dst, const void* src, size_t size)
+    DumpJIT::write(asByteSpan(time));
+    DumpJIT::write(asByteSpan(dst64));
+    DumpJIT::write(asByteSpan(size64));
+    DumpJIT::write(unsafeMakeSpan(static_cast<const uint8_t*>(src), size));
```
Please use `sourceSpan` to avoid the unverified length here.

ddkilzer replied:
Hi @geoffreygaren, you didn't expand the source code enough.  :)

`sourceSpan` is used in the `DumpJIT::write()` method, which is being called here.  The entire `struct DumpJIT` implementation is declared inside this method.

The `src` and `size` in this call come from the `dumpJITMemory()` function arguments on [line 1320](https://github.com/WebKit/WebKit/pull/52678/files#diff-47b7950aaa32998f5ef88ee18bc31dec5970a6f655ca79ec8e02f2ad4570d0a2R1320):
```
void dumpJITMemory(const void* dst, const void* src, size_t size)
```

geoffreygaren reviewed (approved):
This seems fine but what does type-aware malloc do?

Commit:
Fix untyped malloc() calls in JavaScriptCore and WebCore
<https://bugs.webkit.org/show_bug.cgi?id=301118>
<rdar://163059754>

Reviewed by Geoff Garen.

When compiling with type-aware malloc enabled, the compiler has many
hints from which type information may be deduced:
- C-style casts
- reinterpret_cast<>()
- static_cast<>()
- Multiplying a variable by sizeof(TYPE_NAME)

The fix is to find a way to introduce these mechanisms, or to use
WTF_TYPE_AWARE_MALLOC_FUNCTION() when C++ template types aren't
available.

* Source/JavaScriptCore/jit/ExecutableAllocator.cpp:
(JSC::dumpJITMemory):
- Fix untyped malloc() by switching to LazyNeverDestroyed<std::array>.
- Also fix remaining unsafe buffer issues by adopting std::span.

* Source/WTF/wtf/MallocSpan.h:
- Provide specialization for MallocSpan<T, SystemMalloc> that inherits
  from MallocSpan<T, SystemMallocBase<T>> so that <typename T> is
  available when using SystemMallocBase in order to provide type hints
  to malloc() functions via static_cast<T*>().  We use this
  specialization so that other classes used with WTF::MallocSpan are not
  required to adopt a <typename T> template parameter.
* Source/WTF/wtf/PlatformHave.h:
- Change definition of HAVE_TYPE_AWARE_MALLOC and
  WTF_TYPE_AWARE_MALLOC_FUNCTION() to use OS versions.  An enablement
  check is added to SystemMalloc.h where <malloc/malloc.h> may be
  included without affecting preprocesesing of sandbox profiles
  (*.sb.in).
* Source/WTF/wtf/SystemMalloc.h:
- Add enablement checks that _MALLOC_TYPE_ENABLED and
  HAVE_TYPE_AWARE_MALLOC are both enabled at the same time.
- Rename WTF::SystemMalloc to WTF::SystemMallocBase for SystemMalloc
  template specialization in MallocSpan.h (see above).
(WTF::SystemMallocBase::malloc):
(WTF::SystemMallocBase::tryMalloc):
(WTF::SystemMallocBase::zeroedMalloc):
(WTF::SystemMallocBase::tryZeroedMalloc):
(WTF::SystemMallocBase::realloc):
(WTF::SystemMallocBase::tryRealloc):
- Add static_cast<T*>() to provide type hint to malloc() and friends,
  then change the return type from void* to T*.

(WTF::SystemMalloc::malloc): Rename.
(WTF::SystemMalloc::tryMalloc): Rename.
(WTF::SystemMalloc::zeroedMalloc): Rename.
(WTF::SystemMalloc::tryZeroedMalloc): Rename.
(WTF::SystemMalloc::realloc): Rename.
(WTF::SystemMalloc::tryRealloc): Rename.
(WTF::SystemMalloc::free): Rename.
(WTF::SystemMalloc::nextCapacity): Rename.
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::xmlMallocHelper):
(WebCore::xmlFreeHelper):
(WebCore::XMLMalloc::malloc):
- When HAVE(TYPE_AWARE_MALLOC) is enabled, just make XMLMalloc an alias
  of WTF::SystemMalloc since that's effectively what we were doing
  before.  This cleans up typed malloc support for this code.
(WebCore::XMLMalloc::WTF_TYPE_AWARE_MALLOC_FUNCTION): Deleted.

Canonical link: https://commits.webkit.org/302125@main
